src/java/main/org/apache/zookeeper/server/ZooKeeperServerMain.java

```diff
@@ -146,14 +146,26 @@ public void runFromConfig(ServerConfig config) throws IOException, AdminServerEx
             );
             containerManager.start();
 
+            // Watch status of ZooKeeper server. If there is an internal error
+            // then will do a graceful shutdown.
+            while (zkServer.isRunning()) {
```

Comment:
Agreed about the `HealthMonitorThread`, I'd rather not have that. The behavior we are trying to implement is this. ZooKeeper main needs to block until the `isRunning` predicate becomes false. To do it, either we poll as we are doing now in this patch and keep waking up every now and then, or we implement a blocking call in `ZooKeeperServer` that puts all callers to wait until they are notified. We can notify all when `setState` is called and the condition is met. Does it work?